sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -396,8 +396,16 @@ case class TransformWithStateExec(
       new Path(stateSchemaDir, s"${batchId}_${UUID.randomUUID().toString}")
     val metadataPath = new Path(getStateInfo.checkpointLocation, s"${getStateInfo.operatorId}")
     val metadataReader = OperatorStateMetadataReader.createReader(
-      metadataPath, hadoopConf, operatorStateMetadataVersion)
-    val operatorStateMetadata = metadataReader.read()
+      metadataPath, hadoopConf, operatorStateMetadataVersion, batchId)
+    val operatorStateMetadata = try {
+      metadataReader.read()
+    } catch {
+        // If this is the first time we are running the query, there will be no metadata
+        // and this error is expected. In this case, we return None.
+        case ex: Exception if batchId == 0 =>
```

Comment:
Shall we move the `if batchId == 0` before `metadataReader.read` to avoid unnecessary file read?

Reply:
This was intentional. This is basically to guard against the case where we are trying to replay batch 0 but it has the operator metadata already recorded. In this case, we will be able to read that file.

Reply:
qq: but if the file _doesn't_ exist, why don't we just return `None` instead of throwing a `STDS_FAILED_TO_READ_OPERATOR_METADATA`? Do we _need_ to throw?